### PR TITLE
Remove pkg/testutil/assert in favor of testify

### PR DESCRIPTION
**Pull request by aaronlehmann**
I noticed that we're using a homegrown package for assertions. The
functions are extremely similar to testify, but with enough slight
differences to be confusing (for example, Equal takes its arguments in a
different order). We already vendor testify, and it's used in a few
places by tests.

I also found some problems with `pkg/testutil/assert`. For example, the
NotNil function seems to be broken. It checks the argument against
`nil`, which only works for an interface. If you pass in a nil map or
slice, the equality check will fail.

In the interest of avoiding NIH, I'm proposing replacing
`pkg/testutil/assert` with testify. The test code looks almost the same,
but we avoid the confusion of having two similar but slightly different
assertion packages, and having to maintain our own package instead of
using a commonly-used one.

In the process, I found a few places where the tests should halt if an
assertion fails, so I've made those cases (that I noticed) use `require`
instead of `assert`, and I've vendored the `require` package from
testify alongside the already-present `assert` package.

cc @dnephin @vdemeester

### CODE REVIEW

**Comment by dnephin** [comment]
SGTM

>  The functions are extremely similar to testify

By design! So that someday we could do exactly this. At the time `pkg/testutil/assert` was added we weren't using testify yet.

`testify/assert` also has an additional features that makes it easier to use with tabletests. The assertions accept an additional arg that you can pass a string which describes the test. It makes it a lot easier to find the failing case in a table of tests.

**Comment by dnephin** [comment]
I think the only thing missing from testify is a single line `assert.Error()` statement. The testify one only takes an error, so you have to write this:

```go
require.Error(t, err)
assert.Contains(t, err.Error(), expectedError)
```

The problem is that I think this encourages people to skip checking the error message and only check that there is some error.

Maybe we could leave `testutils/assert` for a couple helpers and just change their implementation to use testify?

**Comment by aaronlehmann** [comment]
You can do `require.Error(t, err, expectedError)` with testify. I have this in many places in the PR.

The difference between `testify` and `testutil/assert` is that it checks for an exact match on the error message (which arguably is a good thing to encourage in tests). If you want to check for a substring, it does take an extra line.

I only encountered one place where this came up. If it turns out to be generally useful, I'm fine with having a helper function to do it in a single line (perhaps in `testutil/helpers`?)

**Comment by dnephin** [comment]
Ah, so there is `require.EqualError()`, I might try submitting a PR for `require.ErrorContains()`

**Comment by aaronlehmann** [comment]
Ah, yes, you need to use `EqualError` for that behavior.

Most of the places in this PR unintentionally dropped the error string checking. I'll either add the helper or update tests to check for full error strings where appropriate.

**Comment by vieux** [comment]
SGTM

**Comment by aaronlehmann** [comment]
Updated the PR to restore error substring checking. Will work on converting `cli` tomorrow.

**Comment by vieux** [comment]
@aaronlehmann at least on powerpc, tests are relared

**Comment by aaronlehmann** [comment]
Missed a `git add` :). Will fix tomorrow.

**Comment by aaronlehmann** [comment]
Updated the PR to fully replace `testutil/assert`.

**Comment by unclejack** [comment]
Wasn't testify horribly broken? I recall testify was rejected a while ago as well.

**Comment by unclejack** [comment]
I believe @icecrime has looked into this the last time testify was being discussed.

**Comment by aaronlehmann** [comment]
How is it broken? Why is it better to have our own reimplementation?

**Comment by unclejack** [comment]
@aaronlehmann: I'm not saying we should have our own implementation. It's just that I recall there were objections to using testify. Some issues were brought up related to some bugs in testify. I'll look this up on the tracker and post here.

**Comment by unclejack** [comment]
@aaronlehmann: https://github.com/stretchr/testify/pull/128 was the issue where some problems related to testify were brought up. I haven't used that code in a few years and I haven't checked if it's in a better condition since. My concern was mostly related to avoiding the package(s) if they have bugs/code quality issues. I don't have a strong opinion beyond that.

**Comment by aaronlehmann** [comment]
That PR seems to be a bug fix opened by @icecrime. It was merged more than two years ago.

I don't have a particular attachment to testify, though we use it in swarmkit and haven't encountered any bugs. However, I'm troubled that we've implemented our own `assert` package that's subtly incompatible. I'd strongly recommend we switch to testify unless anyone knows of outstanding bugs or quality issues.

**Comment by vieux** [comment]
as @aaronlehmann said, it seems like the issue in testify has been fixed upstream. I don't see a reason to not merge this one. LGTM

**Review by vdemeester — APPROVED** [review]
LGTM 👼

File: builder/dockerfile/dispatchers_test.go [hunk 1]
```diff
@@ -215,49 +215,49 @@ func TestFromScratch(t *testing.T) {
...
-	assert.NilError(t, err)
-	assert.Equal(t, b.image, expected)
-	assert.Equal(t, b.from.ImageID(), expected)
-	assert.Equal(t, len(b.buildArgs.GetAllAllowed()), 0)
-	assert.Equal(t, len(b.buildArgs.GetAllMeta()), 1)
+	assert.NoError(t, err)
+	assert.Equal(t, expected, b.image)
+	assert.Equal(t, expected, b.from.ImageID())
+	assert.Len(t, b.buildArgs.GetAllAllowed(), 0)
+	assert.Len(t, b.buildArgs.GetAllMeta(), 1)
```
Comment (dnephin) [inline builder/dockerfile/dispatchers_test.go]: I think this is a good example of an inconsistency in the design of testify. 
`assert.Len()`, `assert.Contains()`, and `assert.EqualError()` have actual as the first arg, and expected as the second.

The args for `assert.Equal(t, expected, actual)`  are in the opposite order. They should have been `assert.Equal(t, actual, expected)` to be consistent with the other assertions.

However I don't think that's enough of a reason to keep our own implementation.

**Review by dnephin — APPROVED** [review]
LGTM

**Comment by tonistiigi** [comment]
LGTM

**Pull request merged**